### ISSUE TITLE
Support for custom random generators (cryptographically-secure, seeded, etc)

### ISSUE THREAD

**Issue by skratchdot** [d3/d3-random#6]
I'm wondering if you'd be interested in supporting cryptographically-secure or seeded random generators?

If so, I'd be willing to work on it, but there are multiple approaches, so I'd want your feedback before doing any work.

The idea would be to replace anywhere in the current code base that uses Math.random() with a custom random function. If you don't pass a custom random function, then one would be chosen for you using an algorithm (to be decided by you).

For instance, the `node-uuid` lib uses the following:
- https://github.com/broofa/node-uuid/blob/309512573ec1c60143c257157479a20f7f1f51cd/uuid.js#L7-L70

In node, it tries to use require('crypto').randomBytes ([link](https://nodejs.org/api/crypto.html)), in the browser it tries to WebCrypto ([link](http://www.w3.org/TR/WebCryptoAPI/), [link](https://developer.mozilla.org/en-US/docs/MDN/Doc_status/API/WebCrypto)), then falls back to Math.random() in both scenarios.

I'm not sure how important cryptographically secure generators are in this lib, but it would be cool to have the option, along with seeded generators (that way you could create random distributions that always return the same values- for testing, etc).

Right now, you can use `d3-random` with a seeded generator, but you have to monkey patch Math.random(), i.e.:

```
var d3 = require('d3-random');
var seeded = require('random-seed').create('some custom seed');
Math.random = seeded.random;
// Return a number greater than or equal to 0
// and less than 6, based on the 'some custom seed' seed:
d3.uniform(6)();
```

It would be nice to have a way to use a custom PRNG without monkey patching Math.random(), but that would involve changing or re-thinking the API a bit (I'm guessing).  I'm also not sure how to convert crypto.randomBytes into a good uniform 0..1, so that would require some research (for me at least- I'm sure it's already been solved by others).

As far as the API changes, I have a few ideas, but haven't spent much time thinking about it.  Some quick thoughts:
- Allow passing `random` as either the first or last parameter of all functions.  If `typeof random !== 'function'` then `Math.random` would be used (or a more complex algorithm similar to node-uuid- based on how performant you want this lib to be, and how complex of a "choose default" you want to maintain).
- Allow composing of generators (so you could have a seeded generator that you mix with one of the other ones). I'd need to think about this more, but might be interesting.
- Instead of passing random as an argument, it could be a property of all your functions.  So `uniform.random = Math.random` by default;  Users could then: `var d3 = require('d3-random');d3.uniform.random = myRandomFunction;d3.uniform(6)();`

I have more ideas, but this is getting to be a long post, so I'll wait for feedback :)

More Math.random() / crypto / prng links:
- http://stackoverflow.com/a/5651854
- https://medium.com/@betable/tifu-by-using-math-random-f1c308c4fd9d#.b76ez5ls1

**Comment by skratchdot** [comment]
@mbostock - Any thoughts on this?

**Comment by mbostock** [comment]
I see several issues here:

* I’m not sure whether *cryptographically-secure* random number generators make sense for this library, given that it is intended to generate synthetic distributions of real numbers (floating point) and not cryptography. On the other hand, if the argument is that you simply with a *better* random number generator than Math.random, then sure. I guess what I’m saying is that I think the onus of converting random bytes to a uniform random float in [0, 1) should fall to the user rather than being provided by this library. (A “better” random number generator can also mean a seeded random number generator; related mbostock/d3#2544.)

* If we think a user-substitutable Math.random is a good idea, then I agree we could design a cleaner API than setting the global, although in my experience the monkey-patching hasn’t been too painful. I guess my inclination would be to have the Math.random substitute be an optional argument, always at the end, to all the random number generators. Like so:

```js
var seed = require("random-seed").create("Hello, world!"),
    random = require("d3-random"),
    uniform = random.randomUniform(6, seed.random.bind(seed));
```

Optional arguments are sometimes a little icky to handle. An alternative would be a way of creating a scope, say, a d3.randomSource method which takes a Math.random substitute and returns an object with all the random number generators on it, bound to that source of randomness:

```js
var seed = require("random-seed").create("Hello, world!"),
    random = require("d3-random"),
    uniform = random.randomSource(seed.random.bind(seed)).randomUniform(6);
```

Or equivalently:

```js
var seed = require("random-seed").create("Hello, world!"),
    random = require("d3-random").randomSource(seed.random.bind(seed)),
    uniform = random.randomUniform(6);
```

In other words, `d3_random` is equivalent to `d3_random.randomSource(Math.random)`. I’m not sure which one I prefer, though.

**Comment by skratchdot** [comment]
Yeah.  I was flip-flopping on what is the best way.  An optional argument seems okay, but could get ugly if the APIs want to change.  I like the idea of `d3_random.randomSource(Math.random)` if that can work w/o changing the code too much, or breaking the existing API.  I also don't want to add work, but was hoping to start using `d3-random` in a few projects soon, and didn't want to monkey patch Math.random, so just wanted to get your feedback (before working on a PR that there might not be interest in).

I guess Math.random *could* be monkey patched a bit cleaner using something like `sinon`, but I typically only think of that for use in test-cases.  I'd like to use a seeded generator in a project (along w/ d3-random and no monkey patching).

Anyways, thanks for the feedback!  I'm going to think about this a bit, then potentially have a PR for you at some point.  I agree with you about the cryptographically secure points.  I was more interested in a seeded generator, but had also read a few articles on cryptographically secure generators, so mentioned those as well.

Keep up the great work on all these d3 libs.  I've been enjoying being able to use them as separate/focused packages.

**Comment by mbostock** [comment]
Fixed in 1.1.0.
